### PR TITLE
Documentation for NodeLease feature being promoted to GA

### PR DESCRIPTION
**Pull request by wojtek-t**
Ref https://github.com/kubernetes/enhancements/issues/589

### CODE REVIEW

File: content/en/docs/reference/command-line-tools-reference/feature-gates.md [hunk 1]
```diff
@@ -202,6 +200,9 @@ The following table contains feature gates for graduated or deprecated features.
 | `MountPropagation` | `false` | Alpha | 1.8 | 1.9 |
 | `MountPropagation` | `true` | Beta | 1.10 | 1.11 |
 | `MountPropagation` | `true` | GA | 1.12 | - |
+| `NodeLease` | `false` | Alpha | 1.12 | 1.13 |
+| `NodeLease` | `true` | Beta | 1.14 | 1.16 |
+| `NodeLease` | `true` | Beta | 1.17 | - |
 | `PersistentLocalVolumes` | `false` | Alpha | 1.7 | 1.9 |
```
Comment (kbhawkey) [inline content/en/docs/reference/command-line-tools-reference/feature-gates.md]: hello @wojtek-t .  Some initial suggestions:
- Line 205: Is the state GA?
- Line 177: the `kubelet` or the Kubelet?
- Line 182: This sentence could be reworded and possibly add more explanation, spelling too. Does this make sense?
Compared to the Node resource, the Lease object is lightweight. The Lease resource improves the performance of the node heartbeats as the cluster scales?

From the KEP:
`We will use that object to represent node heartbeat - for each Node there will be a corresponding Lease object with Name equal to Node name in a newly created dedicated namespace `

File: content/en/docs/concepts/architecture/nodes.md [hunk 2]
```diff
@@ -173,19 +173,15 @@ to be unreachable. (The default timeouts are 40s to start reporting
...
-When node lease feature is enabled, each node has an associated `Lease` object in
-`kube-node-lease` namespace that is renewed by the node periodically, and both
-NodeStatus and node lease are treated as heartbeats from the node. Node leases
-are renewed frequently while NodeStatus is reported from node to master only
-when there is some change or enough time has passed (default is 1 minute, which
-is longer than the default timeout of 40 seconds for unreachable nodes). Since
-node lease is much more lightweight than NodeStatus, this feature makes node
-heartbeat significantly cheaper from both scalability and performance
-perspectives.
+Each node has an associated `Lease` object in `kube-node-lease` namespace.
```
Comment (sftim) [inline content/en/docs/concepts/architecture/nodes.md]: nit
```suggestion
Each Node has an associated `Lease` object in the `kube-node-lease` namespace.
```

File: content/en/docs/concepts/architecture/nodes.md [hunk 3]
```diff
@@ -173,19 +173,15 @@ to be unreachable. (The default timeouts are 40s to start reporting
...
+It is periodically renewed by Kubelet and both NodeStatus and node lease
+are treated as heartbeats from the node.
+Node leases are renewed frequently while NodeStatus is reported from node to
+master only where there is some change or enough time has passed (default is
+5 minutes, which is longer than the default timeout of 40 seconds for
+unreachable nodes). Since Lease is much more lighweigh object than Node,
```
Comment (sftim) [inline content/en/docs/concepts/architecture/nodes.md]: 1. Typo: “lighweigh”
2. This page is about Nodes, and the reader could be totally new to Kubernetes. This page is a lot of detail to explain what a Node is. If anything, I'd prefer to slim it further.

Comment (sftim) [reply content/en/docs/concepts/architecture/nodes.md]: How about a heading: “Node leases” plus 100 words on the control plane issuing Node leases to keep track of Node health / failures.

Does NodeStatus even need mentioning here?

File: content/en/docs/concepts/architecture/nodes.md (hunk 2 above)
Comment (sftim) [reply content/en/docs/concepts/architecture/nodes.md]: Alternatively (see other comment) omit this line.

File: content/en/docs/concepts/architecture/nodes.md (hunk 3 above)
Comment (wojtek-t) [reply content/en/docs/concepts/architecture/nodes.md]: I think it does - the reason is that for the last 5 years, NodeStatus was the only signal for heartbeat. Now this is only treated as an additional one.

File: content/en/docs/reference/command-line-tools-reference/feature-gates.md (hunk 1 above)
Comment (wojtek-t) [reply content/en/docs/reference/command-line-tools-reference/feature-gates.md]: rephrased

File: content/en/docs/concepts/architecture/nodes.md (hunk 3 above)
Comment (wojtek-t) [reply content/en/docs/concepts/architecture/nodes.md]: That said, I tried to rephrase it a bit (applied suggestions by @kbhawkey ) and slightly better organize.

File: content/en/docs/concepts/architecture/nodes.md [hunk 4]
```diff
@@ -173,18 +173,19 @@ to be unreachable. (The default timeouts are 40s to start reporting
...
-are renewed frequently while NodeStatus is reported from node to master only
-when there is some change or enough time has passed (default is 1 minute, which
-is longer than the default timeout of 40 seconds for unreachable nodes). Since
-node lease is much more lightweight than NodeStatus, this feature makes node
-heartbeat significantly cheaper from both scalability and performance
-perspectives.
+#### Heartbeats
+
+Each Node  has an associated `Lease` object in `kube-node-lease` namespace.
+It is periodically renewed by the kubelet and both NodeStatus and the Lease
```
Comment (sftim) [inline content/en/docs/concepts/architecture/nodes.md]:
```suggestion
The kubelet updates its Lease object frequently to show that its Node is healthy. In the control plane, both NodeStatus and the Lease
```

File: content/en/docs/concepts/architecture/nodes.md [hunk 5]
```diff
@@ -173,18 +173,19 @@ to be unreachable. (The default timeouts are 40s to start reporting
...
+are treated as heartbeats from the node.
+Node leases are renewed frequently while NodeStatus is reported from node to
```
Comment (sftim) [inline content/en/docs/concepts/architecture/nodes.md]: If accepting the previous suggestion, how about this wording?
```markdown
The kubelet updates NodeStatus either when there is change in status, or
if there has been no update for a configured interval. The default interval
for NodeStatus updates is 5 minutes (much longer than the 40 second
default timeout for unreachable nodes).
```

File: content/en/docs/concepts/architecture/nodes.md [hunk 6]
```diff
@@ -173,18 +173,19 @@ to be unreachable. (The default timeouts are 40s to start reporting
...
+master only where there is some change or enough time has passed (default is
+5 minutes, which is longer than the default timeout of 40 seconds for
+unreachable nodes). Compared to the Node resource, the Lease object is
```
Comment (sftim) [inline content/en/docs/concepts/architecture/nodes.md]: > Compared to the Node resource, the Lease object is lightweight. The Lease resource improves the performance of the node heartbeats as the cluster scales.

As NodeLease is GA (once this is merged, at least), and the oldest supported Kubernetes version will have NodeLease enabled by default, it's less important to explain the older method. I think it's OK to assume that readers are using NodeLease and that it pretty much “just works” for them.
I would cut out these 2 sentences.

File: content/en/docs/concepts/architecture/nodes.md [hunk 7]
```diff
@@ -173,18 +173,19 @@ to be unreachable. (The default timeouts are 40s to start reporting
...
-NodeStatus and node lease are treated as heartbeats from the node. Node leases
-are renewed frequently while NodeStatus is reported from node to master only
-when there is some change or enough time has passed (default is 1 minute, which
-is longer than the default timeout of 40 seconds for unreachable nodes). Since
-node lease is much more lightweight than NodeStatus, this feature makes node
-heartbeat significantly cheaper from both scalability and performance
-perspectives.
+#### Heartbeats
+
+Each Node  has an associated `Lease` object in `kube-node-lease` namespace.
```
Comment (sftim) [inline content/en/docs/concepts/architecture/nodes.md]:
```suggestion
Each Node has an associated Lease object in the `kube-node-lease` {{< glossary_tooltip term_id="namespace" text="namespace">}}.
```

Comment (wojtek-t) [reply content/en/docs/concepts/architecture/nodes.md]: done

File: content/en/docs/concepts/architecture/nodes.md (hunk 5 above)
Comment (wojtek-t) [reply content/en/docs/concepts/architecture/nodes.md]: done

File: content/en/docs/concepts/architecture/nodes.md (hunk 6 above)
Comment (wojtek-t) [reply content/en/docs/concepts/architecture/nodes.md]: I'm not sure - this shows why this is even needed.

File: content/en/docs/concepts/architecture/nodes.md (hunk 4 above)
Comment (wojtek-t) [reply content/en/docs/concepts/architecture/nodes.md]: done

**Comment by wojtek-t** [comment]
PTAL

File: content/en/docs/concepts/architecture/nodes.md [hunk 8]
```diff
@@ -173,18 +173,21 @@ to be unreachable. (The default timeouts are 40s to start reporting
...
-`kube-node-lease` namespace that is renewed by the node periodically, and both
-NodeStatus and node lease are treated as heartbeats from the node. Node leases
-are renewed frequently while NodeStatus is reported from node to master only
-when there is some change or enough time has passed (default is 1 minute, which
-is longer than the default timeout of 40 seconds for unreachable nodes). Since
-node lease is much more lightweight than NodeStatus, this feature makes node
-heartbeat significantly cheaper from both scalability and performance
-perspectives.
+#### Heartbeats
+
```
Comment (kbhawkey) [inline content/en/docs/concepts/architecture/nodes.md]: Hi. I looked at this again. Added a few sentences at the beginning and moved a few more around.
Also created a small list out of the kubelet update info. Not sure about the Lease update interval?
Feel free to use what you like.

Heartbeats, between a node and the control plane, help determine the state of a node.
A heartbeat represents the Node resource `NodeStatus` and the (Lease object) [docs/reference/generated/kubernetes-api/v1.16/#lease-v1-coordination-k8s-io]. Each Node has an associated `Lease` object in the `kube-node-lease` {{< glossary_tooltip term_id="namespace" text="namespace">}}. Compared to the `Node` resource, the `Lease` object is lightweight.

The `kubelet` is responsible for creating and updating the `NodeStatus` and `Lease`.

- The `kubelet` updates the `NodeStatus` either when there is change in status, or
if there has been no update for a configured interval. The default interval
for `NodeStatus` updates is 5 minutes (much longer than the 40 second default timeout for unreachable nodes).
- The `kubelet` creates and updates the `Lease` object. The frequency of the `Lease` updates is independent from the frequency of the `NodeStatus` updates.

With a small footprint heartbeat, the load of maintaining the state of a node is decreased. The `Lease` resource improves the performance of the node heartbeats as the cluster scales.

Comment (wojtek-t) [reply content/en/docs/concepts/architecture/nodes.md]: Thanks - I really like it.
I just modified it slightly - please take a look if that still makes sense to you and suggest improvements.

**Comment by wojtek-t** [comment]
@kbhawkey - thanks a lot for suggestion. PTAL

File: content/en/docs/concepts/architecture/nodes.md [hunk 9]
```diff
@@ -173,18 +173,30 @@ to be unreachable. (The default timeouts are 40s to start reporting
...
+Heatbeats, sent by Kubernetes nodes, help determining their state.
+There are two forms of hearbeats: updates of `NodeStatus` an
```
Comment (kbhawkey) [inline content/en/docs/concepts/architecture/nodes.md]: @wojtek-t , Some more suggestions+nits.
line 171- 174: Somewhat related: these lines seem to be a repeat of lines 187 -190?
line 178: Spelling, wording: Heartbeats, sent by Kubernetes nodes [to the control plane], help determine the health [or availability] of a node.
line 179: Spelling, wording: There are two forms of heartbeats: updates containing `NodeStatus` and updates containing the (Lease object){ ...].
line 181: in the `kube-node-lease` ...
line 191: Wording to make concise: The `kubelet` updates the `Lease` object every 10 seconds (the default update interval).
line 195 - 197: Possibly move content, reduce, and merge with line 183 (@sftim , comments on including this statement?):
`Lease` is a lightweight resource, which improves the performance
of the node heartbeats as the cluster scales.

Comment (wojtek-t) [reply content/en/docs/concepts/architecture/nodes.md]: > line 171- 174: Somewhat related: these lines seem to be a repeat of lines 187 -190?

I disagree - 171-174 is talking what nodecontroller is doing based on signals it get
lines 187-190 is discussing what those signals are

> line 179: Spelling, wording: There are two forms of heartbeats: updates containing NodeStatus and updates containing the (Lease object){ ...].

Those aren't updates that contain something - in the REST api these are updates of objects.


Other comments should addressed.

**Comment by wojtek-t** [comment]
@kbhawkey - PTAL

File: content/en/docs/concepts/architecture/nodes.md (hunk 9 above)
Comment (kbhawkey) [reply content/en/docs/concepts/architecture/nodes.md]: @wojtek-t , OK. Almost there.

line 178: pls. take out square brackets -- I intended you to choose the words that fit best (where are the heartbeats sent? to the control plane, _availability_ vs. _health_):
Heartbeats, sent by Kubernetes nodes, help determine the availability of a node.
line 180: s/hearbeats/heartbeats/
line 180: If this makes sense, replace _an- with _and the_: updates of `NodeStatus` and the ...
line 193: Reverting back to two sentences, as it seems easier to read
The `kubelet` creates and then updates the `Lease` object every 10 seconds (the default update interval).
`Lease` updates occur independently from the `NodeStatus` updates.

Comment (wojtek-t) [reply content/en/docs/concepts/architecture/nodes.md]: Thanks a lot for patience. Done.

**Comment by wojtek-t** [comment]
PTAL

File: content/en/docs/concepts/architecture/nodes.md [hunk 10]
```diff
@@ -173,18 +173,27 @@ to be unreachable. (The default timeouts are 40s to start reporting
...
-is longer than the default timeout of 40 seconds for unreachable nodes). Since
-node lease is much more lightweight than NodeStatus, this feature makes node
-heartbeat significantly cheaper from both scalability and performance
-perspectives.
+#### Heartbeats
+
+Heartbeats, sent by Kubernetes nodes, help determine the availability of a node.
+There are two forms of heartbeats: updates of `NodeStatus` and the
+(Lease object)[docs/reference/generated/kubernetes-api/{{< latest-version >}}/#lease-v1-coordination-k8s-io].
+Each Node has an associated `Lease` object in the `kube-node-lease`
```
Comment (kbhawkey) [inline content/en/docs/concepts/architecture/nodes.md]: Changes seem fine. Looks like I introduced a formatting error. Pls correct line 180:
[Lease object](/docs/reference/generated/kubernetes-api/{{< latest-version >}}/#lease-v1-coordination-k8s-io).

Comment (wojtek-t) [reply content/en/docs/concepts/architecture/nodes.md]: Done - PTAL

Comment (kbhawkey) [reply content/en/docs/concepts/architecture/nodes.md]: /lgtm

**Comment by daminisatya** [comment]
@sftim review from your side? Is it a lgtm?

**Comment by wojtek-t** [comment]
Thanks!
Can someone approve these changes then?

**Comment by sftim** [comment]
@daminisatya ?

File: content/en/docs/concepts/architecture/nodes.md [hunk 11]
```diff
@@ -173,18 +173,27 @@ to be unreachable. (The default timeouts are 40s to start reporting
...
+
+Heartbeats, sent by Kubernetes nodes, help determine the availability of a node.
+There are two forms of heartbeats: updates of `NodeStatus` and the
+[Lease object](/docs/reference/generated/kubernetes-api/{{< latest-version >}}/#lease-v1-coordination-k8s-io).
+Each Node has an associated `Lease` object in the `kube-node-lease`
+{{< glossary_tooltip term_id="namespace" text="namespace">}}.
+Compared to the Node resource, the Lease is a lightweight resource, which improves
+the performance of the node heartbeats as the cluster scales.
+
+The `kubelet` is responsible for creating and updating `NodeStatus` and `Lease`.
```
Comment (sftim) [inline content/en/docs/concepts/architecture/nodes.md]: Late, nit tweaks:
```suggestion
The kubelet is responsible for creating and updating `NodeStatus` and a Lease.
```

File: content/en/docs/concepts/architecture/nodes.md [hunk 12]
```diff
@@ -173,18 +173,27 @@ to be unreachable. (The default timeouts are 40s to start reporting
...
+
+- The `kubelet` updates the `NodeStatus` either when there is change in status,
```
Comment (sftim) [inline content/en/docs/concepts/architecture/nodes.md]:
```suggestion
- The kubelet updates the `NodeStatus` either when there is change in status,
```

File: content/en/docs/concepts/architecture/nodes.md [hunk 13]
```diff
@@ -173,18 +173,27 @@ to be unreachable. (The default timeouts are 40s to start reporting
...
+  or if there has been no update for a configured interval. The default interval
+  for `NodeStatus` updates is 5 minutes (much longer than the 40 second default
+  timeout for unreachable nodes).
+- The `kubelet` creates and then updates the `Lease` object every 10 seconds
```
Comment (sftim) [inline content/en/docs/concepts/architecture/nodes.md]:
```suggestion
- The kubelet creates and then updates its Lease object every 10 seconds
```

File: content/en/docs/concepts/architecture/nodes.md [hunk 14]
```diff
@@ -173,18 +173,27 @@ to be unreachable. (The default timeouts are 40s to start reporting
...
+  (the default update interval). `Lease` updates occur independently from the
```
Comment (sftim) [inline content/en/docs/concepts/architecture/nodes.md]:
```suggestion
(the default update interval). Lease updates occur independently from the
```

File: content/en/docs/concepts/architecture/nodes.md (hunk 12 above)
Comment (wojtek-t) [reply content/en/docs/concepts/architecture/nodes.md]: done

File: content/en/docs/concepts/architecture/nodes.md (hunk 13 above)
Comment (wojtek-t) [reply content/en/docs/concepts/architecture/nodes.md]: done

File: content/en/docs/concepts/architecture/nodes.md (hunk 14 above)
Comment (wojtek-t) [reply content/en/docs/concepts/architecture/nodes.md]: done

File: content/en/docs/concepts/architecture/nodes.md (hunk 11 above)
Comment (wojtek-t) [reply content/en/docs/concepts/architecture/nodes.md]: done

File: content/en/docs/concepts/architecture/nodes.md [hunk 15]
```diff
@@ -173,18 +173,27 @@ to be unreachable. (The default timeouts are 40s to start reporting
...
-is longer than the default timeout of 40 seconds for unreachable nodes). Since
-node lease is much more lightweight than NodeStatus, this feature makes node
-heartbeat significantly cheaper from both scalability and performance
-perspectives.
+#### Heartbeats
+
+Heartbeats, sent by Kubernetes nodes, help determine the availability of a node.
+There are two forms of heartbeats: updates of `NodeStatus` and the
+[Lease object](/docs/reference/generated/kubernetes-api/{{< latest-version >}}/#lease-v1-coordination-k8s-io).
+Each Node has an associated `Lease` object in the `kube-node-lease`
```
Comment (kbhawkey) [inline content/en/docs/concepts/architecture/nodes.md]: These changes look good. If you are cleaning up nits ...
- There is one more Lease to clean up the backticks on line 181.
- Also, I _think_ command tools (kubelet, kubectl) typically are quoted with backticks, throughout the docs.
  For example: https://kubernetes.io/docs/tasks/administer-cluster/out-of-resource/

Comment (sftim) [reply content/en/docs/concepts/architecture/nodes.md]: `kubectl` is 100% a command line tool. `kubelet`, whilst also an executable,  isn't a tool that people _usually_ run in a command line, at least not during typical cluster operation. I see `kubelet` fitting into the same slot as `kube-scheduler`, `kube-controller-manager`, and `kube-proxy`.

(I've used backticks for those names in this very comment but in the docs I typically wouldn't).

Comment (wojtek-t) [reply content/en/docs/concepts/architecture/nodes.md]: Fixed the Lease.

I agree with @sftim that kubelet is in the same category as kube-scheduler, etc.
[in this file kubelet is not in backsticks anywhere]

**Comment by wojtek-t** [comment]
@sftim @kbhawkey - PTAL

File: content/en/docs/concepts/architecture/nodes.md (hunk 15 above)
Comment (kbhawkey) [reply content/en/docs/concepts/architecture/nodes.md]: Hello @wojtek-t . It feels like these changes should be ready 😀 . Okay to override these comments, but if you have the energy for one more change, I'd suggest splitting/adjusting lines 183 - 184.

Lease is a lightweight resource, which improves the performance of the node heartbeats as the cluster scales.
OR
Lease is a smaller resource than the Node, which improves the performance of the node heartbeats as the cluster scales.

Line 186: The kubelet is responsible for creating and updating the `NodeStatus` and a Lease object.

Comment (wojtek-t) [reply content/en/docs/concepts/architecture/nodes.md]: Done.

**Comment by wojtek-t** [comment]
@sftim @kbhawkey - I applied the next comment. PTAL
[That said, can we have it merged? I have a feeling that we can do those iterations indefinitely...]

**Comment by kbhawkey** [comment]
😁 
/approve

**Comment by wojtek-t** [comment]
@kbhawkey - thanks a lot!
